src/bp/ui-admin/src/telemetry/client.ts

```diff
@@ -0,0 +1,146 @@
+import axios from 'axios'
+import { lang } from 'botpress/shared'
+import _ from 'lodash'
+import moment from 'moment'
+import ms from 'ms'
+import uuid from 'uuid'
+
+import store from '../store'
+
+import { EventPackageInfoType, StoreInfoType, TelemetryPackage } from './telemetry_type'
+
+const telemetryPackageVersion = '1.0.0'
+const dataClusterVersion = '1.0.0'
+
+export const axiosConfig = {
+  baseURL: window.TELEMETRY_URL,
+  headers: {
+    withCredentials: false
+  }
+}
+
+const storeInfos: StoreInfoType = {}
+
+const eventPackageInfo: EventPackageInfoType = {}
+
+const toggleLock = (eventName: string) => {
+  if (_.has(eventPackageInfo, eventName)) {
+    eventPackageInfo[eventName].locked = !eventPackageInfo[eventName].locked
+  }
+}
+
+const getEventTimeout = (event: string) => {
+  const item = window.localStorage.getItem(event)
+  if (item !== null) {
+    const timeout = parseInt(item) - new Date().getTime()
+    if (timeout > 0) {
+      return timeout
+    }
+    window.localStorage.removeItem(event)
+  }
+  return 0
+}
+
+const addEventTimeout = (event: string, timeout: number) => {
+  setTimeout(() => toggleLock(event), timeout)
+  window.localStorage.setItem(event, (timeout + moment().valueOf()).toString())
+}
+
+const checkStoreInfoReceived = () => !Object.keys(storeInfos).find(info => _.isEmpty(storeInfos[info]))
+
+export const addStoreInfo = (name: string, pathInStore: string, formatter?: Function) => {
+  storeInfos[name] = {
+    storedInfo: '',
+    loadInfo: () => {
+      const value = _.get(store.getState(), pathInStore)
+      storeInfos[name].storedInfo = formatter ? formatter(value) : value
+    }
+  }
+}
+
+export const getStoreInfo = (name: string) => {
+  if (storeInfos[name]) {
+    return storeInfos[name].storedInfo
+  } else {
+    throw `The information "${name}" asked is not tracked by the telemetry module. Maybe it was not added before it's use?`
+  }
+}
+
+export const addTelemetryEvent = (name: string, timeout: string, getPackage: Function) => {
+  eventPackageInfo[name] = {
+    locked: getEventTimeout(name) >= 0,
+    timeout,
+    getPackage
+  }
+}
+
+const checkTelemetry = async () => {
+  for (const eventName in eventPackageInfo) {
+    if (!eventPackageInfo[eventName].locked) {
+      toggleLock(eventName)
+
+      await sendTelemetry(getTelemetryPackage(eventName, eventPackageInfo[eventName].getPackage()), eventName)
+    }
+  }
+}
+
+export const startTelemetry = () => {
+  addStoreInfo('email', 'user.profile.email')
+
+  addStoreInfo('bp_release', 'version.currentVersion')
+
+  addStoreInfo('bp_license', 'license.licensing.isPro', (info: any) => {
+    return !!info ? 'pro' : 'community'
+  })
```

Comment:
I don't like this, it feels like you made a generic function that adds complexity in the code just to access some object properties

Reply:
This function exists to track the info that every event needs in their packages and to make sure that the values are not undefined. I am changing the code so that there is no need for a lambda function

Reply:
Small changes made in e135fcf to make it clearer